frontend/apps/reader/modules/readerrolling.lua

```diff
@@ -997,7 +997,9 @@ local ENGINE_PROGRESS_INITIAL_DELAY = TimeVal:new{ sec = 2 }
 local ENGINE_PROGRESS_UPDATE_DELAY = TimeVal:new{ usec = 500000 }
 
 function ReaderRolling:showEngineProgress(percent)
-    if G_reader_settings:isFalse("cre_show_progress") then
+    if G_reader_settings and G_reader_settings:isFalse("cre_show_progress") then
+        -- (G_reader_settings might not be available when this is called
```

Comment:
Yes, I mentionned that in https://github.com/koreader/koreader/pull/5406#issuecomment-540059452. It's there, it's used in the main code that sets  the callback - but it's no more accessible to the callback that is run by crengine - but only when in unit tests. It works fine on the emulator, android, kobo...
That "Lua sets come callback to C, that C later uses to call Lua" involves some Lua internal magic (registry, keeping references, Lua stack "environment" that stores globals) that I don't want to dig in to much :) Well, it's somehow clear to me how that works - it's just why it would work differently in Busted/unit tests that I don't want to dig in.
Because may be Busted uses some trick to provides virtuailisation/isolation between tests, that break this single global stuff, etc...

(Moreover, this callback doesn't really need to access G_reader_settings. I just added that in the end to allow disabling the progress meter that slowed things for me when using the emulator over ssh :)

Reply:
> Because may be Busted uses some trick to provides virtuailisation/isolation between tests, that break this single global stuff, etc...

Yes, I very much worked on isolating it better than we did before. There's no virtualization, just a certain degree of environmental isolation. :-) But you should still be within that environment when the callback happens, shouldn't you?

> Moreover, this callback doesn't really need to access G_reader_settings.

Fair enough.

Reply:
> But you should still be within that environment when the callback happens, shouldn't you?

Yes, I _should_ be :) The question is why I am normally, and why I am not when this unit test is run :)

Some possible answers:
a) because busted uses some tricks
b) because we/busted don't really set G_reader_settings as a real real global in that commonrequire - may be global enough to be accessible in the unit tests file - but not global enough to be global for the "detached" C code
c) because for some reason, the callback are called too late and the global is gone (notice these warnings are logged and timestamped at the end of the CircleCI output, dunno why)
d) because I didn't really properly set that callback stuff cause I'm new to that, and I'm just lucky it works in a normal environment :)
e) ...

I sure would like to know the answer - but not to the point of spending much time on it :)